File: apps/mac-shell/Sources/Foundation/Logging/StructuredLogger.swift
import Foundation

enum LogSeverity {
    case debug
    case info
    case warn
    case error
}

struct LogRecord<TPayload: Codable & Equatable>: Equatable {
    let severity: LogSeverity
    let category: String
    let event: BridgeEventEnvelope<TPayload>
}

protocol LogSink {
    associatedtype Payload: Codable & Equatable

    func append(_ record: LogRecord<Payload>)
    func records() -> [LogRecord<Payload>]
}

final class InMemoryLogSink<TPayload: Codable & Equatable>: LogSink {
    typealias Payload = TPayload

    private var entries: [LogRecord<TPayload>] = []

    func append(_ record: LogRecord<TPayload>) {
        entries.append(record)
    }

    func records() -> [LogRecord<TPayload>] {
        entries
    }
}

final class StructuredLogger<TPayload: Codable & Equatable> {
    private let sink: InMemoryLogSink<TPayload>

    init(sink: InMemoryLogSink<TPayload>) {
        self.sink = sink
    }

    func logEvent(_ severity: LogSeverity, event: BridgeEventEnvelope<TPayload>) {
        sink.append(LogRecord(
            severity: severity,
            category: "bridge_event",
            event: event
        ))
    }

    func records() -> [LogRecord<TPayload>] {
        sink.records()
    }
}

File: apps/mobile/lib/foundation/logging/structured_logger.dart
import '../contracts/bridge_contracts.dart';

enum LogSeverity { debug, info, warn, error }

class LogRecord<TPayload> {
  const LogRecord({
    required this.severity,
    required this.category,
    required this.event,
  });

  final LogSeverity severity;
  final String category;
  final BridgeEventEnvelope<TPayload> event;
}

abstract class LogSink<TPayload> {
  void append(LogRecord<TPayload> record);
  List<LogRecord<TPayload>> records();
}

class InMemoryLogSink<TPayload> implements LogSink<TPayload> {
  final List<LogRecord<TPayload>> _records = <LogRecord<TPayload>>[];

  @override
  void append(LogRecord<TPayload> record) {
    _records.add(record);
  }

  @override
  List<LogRecord<TPayload>> records() =>
      List<LogRecord<TPayload>>.unmodifiable(_records);
}

class StructuredLogger<TPayload> {
  StructuredLogger(this._sink);

  final LogSink<TPayload> _sink;

  void logEvent(LogSeverity severity, BridgeEventEnvelope<TPayload> event) {
    _sink.append(
      LogRecord<TPayload>(
        severity: severity,
        category: 'bridge_event',
        event: event,
      ),
    );
  }

  void logSecurityAudit({
    required LogSeverity severity,
    required String eventId,
    required String threadId,
    required String occurredAt,
    required SecurityAuditEventDto auditEvent,
    required TPayload Function(SecurityAuditEventDto event) payloadMapper,
  }) {
    logEvent(
      severity,
      BridgeEventEnvelope<TPayload>(
        contractVersion: contractVersion,
        eventId: eventId,
        threadId: threadId,
        kind: BridgeEventKind.securityAudit,
        occurredAt: occurredAt,
        payload: payloadMapper(auditEvent),
      ),
    );
  }

  List<LogRecord<TPayload>> records() => _sink.records();
}
